db.go

```diff
@@ -92,6 +92,43 @@ func UnmarshalRaw(rawData, v interface{}) (ok bool, err error) {
 	return true, nil
 }
 
+// Used for RawQuery Unmarshaling
+type RawQuery[I any] struct {
+	Status string `json:"status"`
+	Time   string `json:"time"`
+	Result I      `json:"result"`
+	Detail string `json:"detail"`
+}
+
+// SmartUnmarshal using generics for return desired type.
+// Supports both raw and normal queries.
+func SmartUnmarshal[I any](respond interface{}, wrapperError error) (data I, err error) {
```

Comment:
I wasn't familiar with this pattern where a returned tuple can be used as an input tuple to a function! Thanks for demonstrating.
That said, I am not sure if we would want such a design. We can have helper functions that provide this logic, but I think the `SmartUnmarshal` should itself be `(interface{}) -> (I, error)`. Then you could have `SmartUnmarshalWithQuery` that follows this signature. What do you think?
What I am trying to achieve with this feedback is thinking how would this code live
- will this be a long term alternative to Unmarshal? Preferrable? Less preferrable (considering potential golang frameworks)?
- will people find the signature intuitive?
- is this signature more or less convenient? We can see from the tests that it is very convenient there. But is that how people will be accessing this function, or will they be writing retry logic and retriable/nonretriable error handling before such deserialisation?


Would be interested to hear your (and others) thoughts around this.

Reply:
Hey, firstly I was thinking to create another function like said can take `interface{}` 
but when I used the library the biggest struggles I had was

- there was so much error checking needed, this was making code longer for no reason.
- there were two separate unmarshal, I needed one can handle both
- the  `interface{}`  that Query returns, I was passing interface to function to function for no reason.
- Generics usage was hard with raw Queries.
- I wanted to type safety like I wanted to specify what will be returned.

That why I created [SmartQuery](https://github.com/surrealdb/surrealdb.go/pull/42#top) this on mind

- One-line expression
- Will handle errors and return if there was an error.
- Will handle both Queries
- Will support generic out-of-the-box.
- Will return Type.
- Will wrap other Query

That's why _SmartUnmarshal_ input get `respond interface{}, wrapperError error` for convenience.
also, it is much cleaner with one-line expressions.


Ex.

```go
// What needed now
func Getstaff(plyarr *[]Player) (err error) {
	data, err := db.Query("select * from users where Staff != null", nil)
	if err != nil {
		return err
	}
	ok, err := surrealdb.UnmarshalRaw(data, plyarr)
	if err != nil {
		return err
	}
	if !ok {
		err = errors.New("Not Ok")
	}
}
```


```go
// After PR
func Getstaff(plyarr *[]Player) (err error) {
	plyarr, err = surrealdb.SmartUnmarshal[*[]IPlayer](db.Query("select * from users where Staff != null", nil))
}
```

it makes a difference huge when you have a code base that struct has methods for delete, gets, update, and init.
also, it can be used with Generics functions for just use one get to function, etc.

Reply:
This can provide us efficiency.

Reply:
Yeah, actually thinking about it more I am starting to agree with you. I was going to say that across drivers we need to be versatile and provide a common baseline of behaviour that is consistent. But that is a lot of overhead for what we have today - your changes are more welcome and significant than such standardisation. So happy to approve and merge. Thanks @ElecTwix !